File: scripts/upscale_assets.py
#!/usr/bin/env python3
"""
upscale_assets.py — 3× nearest-neighbor upscale pipeline for Ninja Adventure assets.

Scales all PNG files from 16×16px (Ninja Adventure source) to 48×48px
(Cloud Quest CONFIG.TILE_SIZE = 48). Uses nearest-neighbor interpolation
only — no anti-aliasing, no bilinear smoothing. Preserves alpha channel.
Strips embedded ICC profiles from source PNGs to avoid colour-space warnings.

Usage:
    python scripts/upscale_assets.py \\
        --input  ./ninja-adventure-raw/ \\
        --output ./assets/ \\
        --scale  3

The input tree is walked recursively. Each PNG is written to a mirrored
subdirectory under --output with the same relative path.

Asset categories are inferred from the first path component under --input:

    characters/   monsters/   tiles/   items/   effects/   ui/

Any directory name not in that list is counted under "other".

Exit codes:
    0 — success (at least one file processed)
    1 — invalid or missing command-line arguments, or no PNG files found under --input
    2 — import error (Pillow not installed)
"""

import argparse
import os
import sys

try:
    from PIL import Image
except ImportError:
    print(
        "ERROR: Pillow is not installed.\n"
        "Install it with:  pip install Pillow",
        file=sys.stderr,
    )
    sys.exit(2)

KNOWN_CATEGORIES = {"characters", "monsters", "tiles", "items", "effects", "ui"}


def upscale_png(src_path: str, dst_path: str, scale: int) -> None:
    """Upscale a single PNG with nearest-neighbor interpolation.

    Strips any embedded ICC profile so output PNGs are profile-free.
    Preserves the full alpha channel (RGBA).
    """
    with Image.open(src_path) as img:
        # Convert to RGBA to normalise mode and preserve transparency.
        if img.mode != "RGBA":
            img = img.convert("RGBA")
        # Explicitly strip any ICC profile — resize() propagates img.info,
        # so we must clear it before scaling even if the image is already RGBA.
        img.info.pop("icc_profile", None)

        new_w = img.width * scale
        new_h = img.height * scale
        upscaled = img.resize((new_w, new_h), Image.NEAREST)

    os.makedirs(os.path.dirname(dst_path), exist_ok=True)
    # Save without icc_profile key — Pillow will not embed one
    upscaled.save(dst_path, format="PNG", optimize=False)


def infer_category(rel_path: str) -> str:
    """Return the asset category from the first component of a relative path."""
    parts = rel_path.replace("\\", "/").split("/")
    first = parts[0].lower() if parts else "other"
    return first if first in KNOWN_CATEGORIES else "other"


def process_directory(input_dir: str, output_dir: str, scale: int) -> dict:
    """Walk input_dir, upscale every PNG, write to output_dir.

    Returns a manifest dict:  { category: count }
    """
    manifest: dict[str, int] = {}
    found = False

    for dirpath, _dirnames, filenames in os.walk(input_dir):
        for filename in filenames:
            if not filename.lower().endswith(".png"):
                continue

            found = True
            src_path = os.path.join(dirpath, filename)
            rel_path = os.path.relpath(src_path, input_dir)
            dst_path = os.path.join(output_dir, rel_path)

            category = infer_category(rel_path)
            manifest[category] = manifest.get(category, 0) + 1

            upscale_png(src_path, dst_path, scale)
            print(f"  {rel_path}")

    if not found:
        return {}

    return manifest


def main() -> None:
    parser = argparse.ArgumentParser(
        description="3× nearest-neighbor upscale pipeline for Ninja Adventure assets.",
    )
    parser.add_argument(
        "--input",
        required=True,
        metavar="DIR",
        help="Root directory of raw Ninja Adventure assets.",
    )
    parser.add_argument(
        "--output",
        required=True,
        metavar="DIR",
        help="Root output directory (assets/ subfolders are created automatically).",
    )
    parser.add_argument(
        "--scale",
        type=int,
        default=3,
        metavar="N",
        help="Integer scale factor (default: 3 → 16px → 48px).",
    )
    args = parser.parse_args()

    input_dir = os.path.abspath(args.input)
    output_dir = os.path.abspath(args.output)
    scale = args.scale

    if not os.path.isdir(input_dir):
        print(f"ERROR: input directory not found: {input_dir}", file=sys.stderr)
        sys.exit(1)

    if scale < 1:
        print(f"ERROR: --scale must be a positive integer, got {scale}", file=sys.stderr)
        sys.exit(1)

    print(f"Upscaling assets {scale}× (nearest-neighbor)")
    print(f"  Input  : {input_dir}")
    print(f"  Output : {output_dir}")
    print()

    manifest = process_directory(input_dir, output_dir, scale)

    if not manifest:
        print("ERROR: no PNG files found under input directory.", file=sys.stderr)
        sys.exit(1)

    total = sum(manifest.values())
    print()
    print("── Manifest ──────────────────────────────")
    for category in sorted(manifest):
        print(f"  {category:<16} {manifest[category]:>4} file(s)")
    print(f"  {'TOTAL':<16} {total:>4} file(s)")
    print("──────────────────────────────────────────")
    print(f"\nDone — {total} PNG(s) upscaled {scale}× → written to {output_dir}")


if __name__ == "__main__":
    main()

File: scripts/upscale-assets.js
/**
 * scripts/upscale-assets.js
 *
 * Batch 3× nearest-neighbor upscale for Ninja Adventure assets (16→48px).
 *
 * Ninja Adventure uses 16×16px tiles. Cloud Quest uses CONFIG.TILE_SIZE = 48px.
 * 16 × 3 = 48 — clean integer scale, zero fractional artifacts.
 *
 * Requires ImageMagick 6 (`convert`) or ImageMagick 7 (`magick`) to be installed.
 *
 * Usage:
 *   node scripts/upscale-assets.js --input <path> --output <path> [--scale 3]
 *
 * Example — upscale the full Ninja Adventure pack into the assets folder:
 *   node scripts/upscale-assets.js \
 *     --input /tmp/NinjaAdventure \
 *     --output assets/ninja-adventure-upscaled
 *
 * The output directory tree mirrors the input tree.
 * Only .png files are processed; all other files are copied verbatim.
 * Existing output files are skipped unless --force is passed.
 *
 * Exit codes:
 *   0  — success
 *   1  — ImageMagick not found or fatal error
 */

import { execSync }              from 'node:child_process'
import { fileURLToPath }         from 'node:url'
import path                      from 'node:path'
import fs                        from 'node:fs'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const ROOT      = path.resolve(__dirname, '..')

// ---------------------------------------------------------------------------
// CLI parsing
// ---------------------------------------------------------------------------

function parseArgs(argv) {
  const args = { input: null, output: null, scale: 3, force: false }
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === '--input'  && argv[i + 1]) { args.input  = argv[++i]; continue }
    if (argv[i] === '--output' && argv[i + 1]) { args.output = argv[++i]; continue }
    if (argv[i] === '--scale'  && argv[i + 1]) { args.scale  = Number(argv[++i]); continue }
    if (argv[i] === '--force') { args.force = true; continue }
  }
  return args
}

const args = parseArgs(process.argv)

if (!args.input || !args.output) {
  console.error('Usage: node scripts/upscale-assets.js --input <src> --output <dest> [--scale 3] [--force]')
  process.exit(1)
}

const SCALE = args.scale
if (!Number.isFinite(SCALE) || SCALE <= 0 || !Number.isInteger(SCALE)) {
  console.error(`--scale must be a positive integer (got: ${args.scale})`)
  process.exit(1)
}

const INPUT  = path.resolve(ROOT, args.input)
const OUTPUT = path.resolve(ROOT, args.output)

if (!fs.existsSync(INPUT)) {
  console.error(`Input directory not found: ${INPUT}`)
  process.exit(1)
}

if (!fs.statSync(INPUT).isDirectory()) {
  console.error(`Input path is not a directory: ${INPUT}`)
  process.exit(1)
}

if (fs.existsSync(OUTPUT) && !fs.statSync(OUTPUT).isDirectory()) {
  console.error(`Output path exists but is not a directory: ${OUTPUT}`)
  process.exit(1)
}

function isSameOrNestedPath(parentPath, childPath) {
  const rel = path.relative(parentPath, childPath)
  return rel === '' || (!rel.startsWith('..') && !path.isAbsolute(rel))
}

if (isSameOrNestedPath(INPUT, OUTPUT)) {
  console.error(
    'Output directory must not be the same as or inside the input directory.\n' +
    `  input:  ${INPUT}\n` +
    `  output: ${OUTPUT}`
  )
  process.exit(1)
}

// ---------------------------------------------------------------------------
// Detect ImageMagick
// ---------------------------------------------------------------------------

function detectImageMagick() {
  // Prefer ImageMagick 7 (`magick`), fall back to ImageMagick 6 (`convert`).
  for (const cmd of ['magick', 'convert']) {
    try {
      execSync(`${cmd} --version`, { stdio: 'pipe' })
      return cmd
    } catch {
      // not available, try next
    }
  }
  return null
}

const MAGICK = detectImageMagick()
if (!MAGICK) {
  console.error(
    'ImageMagick not found.\n' +
    '  Install on Ubuntu/Debian: sudo apt install imagemagick\n' +
    '  Install on macOS:          brew install imagemagick\n' +
    '  Install on Windows:        https://imagemagick.org/script/download.php#windows'
  )
  process.exit(1)
}

// ---------------------------------------------------------------------------
// Walk + process
// ---------------------------------------------------------------------------

let processed = 0
let skipped   = 0
let copied    = 0
let errors    = 0

function walk(dir) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const srcPath = path.join(dir, entry.name)
    const relPath = path.relative(INPUT, srcPath)
    const dstPath = path.join(OUTPUT, relPath)

    if (entry.isDirectory()) {
      fs.mkdirSync(dstPath, { recursive: true })
      walk(srcPath)
      continue
    }

    if (!entry.isFile()) continue

    const ext = path.extname(entry.name).toLowerCase()

    if (ext === '.png') {
      upscalePng(srcPath, dstPath, relPath)
    } else {
      copyFile(srcPath, dstPath, relPath)
    }
  }
}

function upscalePng(src, dst, rel) {
  if (!args.force && fs.existsSync(dst)) {
    console.log(`  skip  ${rel}`)
    skipped++
    return
  }

  fs.mkdirSync(path.dirname(dst), { recursive: true })

  try {
    // -filter Point        — nearest-neighbor resampling
    // -resize ${SCALE*100}%  — scale by SCALE×
    // +profile '*'         — strip colour profiles to keep file size small
    execSync(
      `${MAGICK} "${src}" -filter Point -resize ${SCALE * 100}% +profile '*' "${dst}"`,
      { stdio: 'pipe' }
    )
    console.log(`  up${SCALE}×  ${rel}`)
    processed++
  } catch (err) {
    console.error(`  ERROR ${rel}: ${err.stderr?.toString().trim() ?? err.message}`)
    errors++
  }
}

function copyFile(src, dst, rel) {
  if (!args.force && fs.existsSync(dst)) {
    skipped++
    return
  }
  fs.mkdirSync(path.dirname(dst), { recursive: true })
  fs.copyFileSync(src, dst)
  console.log(`  copy  ${rel}`)
  copied++
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

console.log(`Upscaling Ninja Adventure assets ${SCALE}× (${16 * SCALE}px output)`)
console.log(`  Input:  ${INPUT}`)
console.log(`  Output: ${OUTPUT}`)
console.log(`  Tool:   ${MAGICK}`)
console.log()

fs.mkdirSync(OUTPUT, { recursive: true })
walk(INPUT)

console.log()
console.log(`Done — ${processed} upscaled, ${copied} copied, ${skipped} skipped, ${errors} errors`)

if (errors > 0) process.exit(1)
